mqtt/mqtt-bridge/src/pump/mod.rs

```diff
@@ -0,0 +1,314 @@
+mod builder;
+
+pub use builder::Builder;
+
+// #![allow(dead_code, unused_imports, unused_variables)]
+use futures_util::{
+    stream::{StreamExt, TryStreamExt},
+    FutureExt,
+};
+use tokio::{
+    select,
+    sync::{
+        mpsc::{self, error::SendError},
+        oneshot,
+    },
+};
+use tracing::{debug, error, info};
+
+use mqtt3::PublishHandle;
+
+use crate::{
+    bridge::BridgeError,
+    client::{ClientShutdownHandle, EventHandler, MqttClient},
+    connectivity::ConnectivityState,
+    persist::{MessageLoader, PublicationStore, WakingMemoryStore},
+    settings::ConnectionSettings,
+    upstream::{CommandId, RpcCommand},
+};
+
+#[derive(Debug, PartialEq)]
+pub enum PumpMessage {
+    ConnectivityUpdate(ConnectivityState),
+    ConfigurationUpdate(ConnectionSettings),
+    RpcCommand(CommandId, RpcCommand),
+    RpcAck(CommandId, RpcCommand),
+    Shutdown,
+}
+
+pub struct PumpHandle {
+    sender: mpsc::Sender<PumpMessage>,
+}
+
+impl PumpHandle {
+    fn new(sender: mpsc::Sender<PumpMessage>) -> Self {
+        Self { sender }
+    }
+
+    pub async fn send(&mut self, message: PumpMessage) -> Result<(), PumpError> {
+        self.sender.send(message).await.map_err(PumpError)
+    }
+}
+
+#[cfg(test)]
+pub fn channel() -> (PumpHandle, mpsc::Receiver<PumpMessage>) {
+    let (tx, rx) = tokio::sync::mpsc::channel(10);
+    (PumpHandle::new(tx), rx)
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct PumpError(SendError<PumpMessage>);
+
+/// Pump used to connect to either local broker or remote brokers (including the upstream edge device)
+/// It contains an mqtt client that connects to a local/remote broker
+/// After connection there are two simultaneous processes:
+/// 1) persist incoming messages into an ingress store to be used by another pump
+/// 2) publish outgoing messages from an egress store to the local/remote broker
+pub struct Pump<H> {
+    subscriptions: Vec<String>,
+    messages_send: mpsc::Sender<PumpMessage>,
+    messages: MessagesProcessor,
+    egress: Egress,
+    ingress: Ingress<H>,
+}
+
+impl<H: EventHandler> Pump<H> {
+    fn new(
+        messages_send: mpsc::Sender<PumpMessage>,
+        messages_recv: mpsc::Receiver<PumpMessage>,
+        client: MqttClient<H>,
+        subscriptions: Vec<String>,
+        loader: MessageLoader<WakingMemoryStore>,
+        store: PublicationStore<WakingMemoryStore>,
+    ) -> Result<Self, BridgeError> {
+        let client_shutdown = client.shutdown_handle()?;
+        let publish_handle = client
+            .publish_handle()
+            .map_err(BridgeError::PublishHandle)?;
+
+        let egress = Egress::new(publish_handle, loader, store);
+        let ingress = Ingress::new(client, client_shutdown);
```

Comment:
I think ingress client should have a list of EventHandlers and process the messages, one of the handlers will be the connectivity

Reply:
It's going to be in MessagesProcessor

Reply:
Not clear why in message processor, I though that's used for messages between pumps. The eventhandler is to process messages from ingress pumo

Reply:
nvm, I think I missed last change, I think it makes sense